### PR TITLE
Feature/dd20 format validation

### PR DESCRIPTION
**Pull request by cjjohansen**
**adds check for format changes in dd20 files**

This is a suggestion. Feel free to suggest better placement of code 

- we should move hard coded settings out into  environment variables
- consider improved structure for settings/configuration. 
- consider  two groups of configuration one for line data and one for station data
- consider iterate over  dd20 data sheets  so 3rd type could theroretically be included.

Category:

- [x ] New Feature
- [ ] Bugfix
- [ ] Documentation update

Checklist:

- [ ] I have bumped all relevant version numbers
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have documented my changes if this is needed
- [ ] I have described my changes in the "description of your changes" field
- [ ] I have tested this change on a running cluster 
- [ ] I have created/corrected automated tests related to the changes, if possible/necessary
- [ ] I have checked that new and existing unit tests pass locally with my changes
- [ ] I have squashed commits if necessary

### CODE REVIEW

File: app/helpers/dd20_format_validation.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import pandas as pd
+import hashlib
+
+
+def validate_dd20_format(data_frame: pd.DataFrame, expected_hash: int) -> bool:
+    return calculate_dd20_format_hash(data_frame) == expected_hash
```
Comment (CodeWitten) [inline app/helpers/dd20_format_validation.py:6]: add docstring.

File: app/helpers/parse_dd20.py [hunk 2]
```diff
@@ -857,12 +860,26 @@ def parse_dd20_excelsheets_to_dataframe(
...
+        raise dd20_format_validation.DD20FormatError(error_message) 
+
     # Instantiation of objects for parsing data from station and line sheets of DD20
     data_station = DD20StationDataframeParser(
         df_station=dd20_dataframe_dict[sheetname_stationsdata]
     )
+
+    if not dd20_format_validation.validate_dd20_format(dd20_dataframe_dict[sheetname_linedata], line_data_valid_format_hash_value):
+        error_message = f"invalid dd20 file format detected for {sheetname_linedata}"
+        log.critical(error_message) # toto improve error message
```
Comment (CodeWitten) [inline app/helpers/parse_dd20.py:876]: i suggest we just raise the error and the log it when we call the function. At least this is the method used for this repo for now. Therefore i suggest removing the log line.

File: app/helpers/parse_dd20.py [hunk 3]
```diff
@@ -857,12 +860,26 @@ def parse_dd20_excelsheets_to_dataframe(
         header=header_index,
     )
 
+
+    if not dd20_format_validation.validate_dd20_format(dd20_dataframe_dict[sheetname_stationsdata], station_data_valid_format_hash_value):
+        error_message = f"invalid dd20 file format detected for {sheetname_stationsdata}"
+        log.critical(error_message) # toto improve error message
```
Comment (CodeWitten) [inline app/helpers/parse_dd20.py:866]: i suggest we just raise the error and the log it when we call the function. At least this is the method used for this repo for now. Therefore i suggest removing the log line.

File: tests/test_dd20_format_validation.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        sheet_name=sheets,
+        header=header_index,
+    )
+    return dd20_dataframes
+
+def test_can_validate_dd20_line_data_summer_format(dd20_data_frames):
+
+    expected_hash = "9cf51349b6b13d3c52deb66bf569eb49"
+
+    dd20_line_data_summer_data_frame = dd20_data_frames["Linjedata - Sommer"]
```
Comment (CodeWitten) [inline tests/test_dd20_format_validation.py:36]: hardcoding of sheet name. Could we make variable for sheet names, to only have them configures on one place?

File: tests/test_dd20_format_validation.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    assert validators.validate_dd20_format(
+        dd20_line_data_summer_data_frame, expected_hash
+    )
+
+def test_can_validate_dd20_line_data_station_format(dd20_data_frames):
+
+    expected_hash = "f06edffbc927aea71f0a501f520cf583"
+
+    dd20_line_data_station_data_frame = dd20_data_frames["Stationsdata"]
```
Comment (CodeWitten) [inline tests/test_dd20_format_validation.py:50]: hardcoding of sheet name. Could we make variable for sheet names, to only have them configures on one place?

File: app/helpers/parse_dd20.py [hunk 6]
```diff
@@ -829,6 +830,8 @@ def parse_dd20_excelsheets_to_dataframe(
     header_index: int = 1,
     sheetname_linedata: str = "Linjedata - Sommer",
     sheetname_stationsdata: str = "Stationsdata",
+    line_data_valid_format_hash_value = "9cf51349b6b13d3c52deb66bf569eb49", # todo refactor and move to env variables
+    station_data_valid_format_hash_value = "f06edffbc927aea71f0a501f520cf583" # todo refactor and move to env variables
```
Comment (CodeWitten) [inline app/helpers/parse_dd20.py:834]: hardcoding of sheet name. Could we make variable for sheet names, to only have them configures on one place?

Comment (cjjohansen) [reply app/helpers/parse_dd20.py:834]: comment in wrong place.

File: tests/test_dd20_format_validation.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+"""
+Tests for validating dd20 format by calculating a checksum / hash value
+"""
+
+# from pickle import TRUE
```
[the remark this answers is not in the document]
Comment (cjjohansen) [reply tests/test_dd20_format_validation.py:5]: done

File: app/helpers/parse_dd20.py (hunk 2 above)
Comment (cjjohansen) [reply app/helpers/parse_dd20.py:876]: removed log line

File: app/helpers/parse_dd20.py (hunk 3 above)
Comment (cjjohansen) [reply app/helpers/parse_dd20.py:866]: removed log line

File: tests/test_dd20_format_validation.py (hunk 4 above)
Comment (cjjohansen) [reply tests/test_dd20_format_validation.py:36]: fixed

File: tests/test_dd20_format_validation.py (hunk 5 above)
Comment (cjjohansen) [reply tests/test_dd20_format_validation.py:50]: fixed

File: app/helpers/dd20_format_validation.py (hunk 1 above)
Comment (cjjohansen) [reply app/helpers/dd20_format_validation.py:6]: Done

**Review by CodeWitten — APPROVED** [review]

**Comment by CodeWitten** [comment]
@cjjohansen hash lib skal

**Comment by CodeWitten** [comment]
@cjjohansen der skal forresten lige fjernes en række linting ting, glemte jeg at nævne. Du kan se linting beskeder her:
https://github.com/energinet-singularity/conductor-data-provider/runs/6706970967?check_suite_focus=true

De kan sikker fikses med auto format i vscode.

**Comment by swedchef** [comment]
This has been superseded by other branch.

**Pull request closed without merging**